src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Analyzer.cs

```diff
@@ -0,0 +1,267 @@
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                IInvocationOperation invocation = (IInvocationOperation)context.Operation;
+
+                if (!(context.ContainingSymbol is IMethodSymbol containingSymbol))
+                {
+                    return;
+                }
+
+                if (!ShouldDiagnose(
+                    invocation,
+                    containingSymbol,
+                    cancellationTokenType,
+                    out string? cancellationTokenParameterName))
+                {
+                    return;
+                }
+
+                // Only underline the method name, not the whole invocation
+                SyntaxNode? expressionNode = GetMethodNameNode(context.Operation.Syntax);
+                if (expressionNode != null)
+                {
+                    context.ReportDiagnostic(expressionNode.CreateDiagnostic(ForwardCancellationTokenToInvocationsRule, cancellationTokenParameterName, invocation.TargetMethod.Name));
+                }
+            },
+            OperationKind.Invocation);
+        }
+
+        // Determines if an invocation should trigger a diagnostic for this rule or not.
+        private static bool ShouldDiagnose(
+            IInvocationOperation invocation,
+            IMethodSymbol containingSymbol,
+            INamedTypeSymbol cancellationTokenType,
+            [NotNullWhen(returnValue: true)] out string? cancellationTokenParameterName)
+        {
+            cancellationTokenParameterName = null;
+
+            IMethodSymbol method = invocation.TargetMethod;
+
+            // Verify that the current invocation is not passing a token already
+            if (invocation.Arguments.Any(a => a.Parameter.Type.Equals(cancellationTokenType) && !a.IsImplicit))
+            {
+                return false;
+            }
+
+            // Check if the invocation's method has either an optional implicit ct not being used or a params ct parameter not being used or an overload that takes a ct
+            if (!InvocationMethodTakesAToken(method, invocation.Arguments, cancellationTokenType) &&
+                !MethodHasCancellationTokenOverload(method, cancellationTokenType))
+            {
+                return false;
+            }
+
+            // Check if the ancestor method has a ct that we can pass to the invocation
+            if (!VerifyAncestorOnlyHasOneCancellationTokenAsLastArgument(cancellationTokenType, GetContainingSymbol(invocation, containingSymbol), out cancellationTokenParameterName))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        // Try to find the most immediate containing symbol (anonymous or local function). If none is found, return the context containing symbol.
+        private static IMethodSymbol GetContainingSymbol(IInvocationOperation invocation, IMethodSymbol containingSymbol)
```

Comment:
Can we make this an extension method and move it to `IOperationExtensions.cs`?

```suggestion
        /// <summary>
        /// Gets the immediate containing lambda or local function, if any. Otherwise, returns the fallback containing member symbol.
        /// </summary>
        internal static ISymbol GetImmediateContainingSymbol(this IOperation invocation, ISymbol fallbackContainingSymbol)
```

Reply:
If we decide to implement the suggestion from https://github.com/dotnet/roslyn-analyzers/pull/3641/files#r436914521, then it might make sense to retain this as a private helper here and change it to `(IOperation lambdaOrLocalFunctionOperation, IMethodSymbol lambdaOrLocalFunctionMethod))? GetContainingLambdaOrLocalFunction(IOperation operation);` so we can loop through all the containing methods.